File: cocos/renderer/core/program-lib.ts
/*
 Copyright (c) 2017-2018 Xiamen Yaji Software Co., Ltd.

 http://www.cocos.com

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated engine source code (the "Software"), a limited,
  worldwide, royalty-free, non-assignable, revocable and non-exclusive license
 to use Cocos Creator solely to develop games on your target platforms. You shall
  not use Cocos Creator software for developing other software or tools that's
  used for developing games. You are not granted to publish, distribute,
  sublicense, and/or sell copies of Cocos Creator.

 The software or tools in this License Agreement are licensed, not sold.
 Xiamen Yaji Software Co., Ltd. reserves all rights not expressly granted to you.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 THE SOFTWARE.
*/

/**
 * @category material
 */

import { IBuiltinInfo, IDefineInfo, IShaderInfo } from '../../3d/assets/effect-asset';
import { GFXBindingType, GFXShaderType } from '../../gfx/define';
import { GFXAPI, GFXDevice } from '../../gfx/device';
import { GFXShader } from '../../gfx/shader';
import { IInternalBindingDesc, localBindingsDesc } from '../../pipeline/define';
import { RenderPipeline } from '../../pipeline/render-pipeline';
import { IDefineMap } from './pass';

let _shdID = 0;
interface IDefineRecord extends IDefineInfo {
    _map: (value: any) => number;
    _offset: number;
}
interface IProgramInfo extends IShaderInfo {
    id: number;
    defines: IDefineRecord[];
    globalsInited: boolean;
    localsInited: boolean;
}
export interface IDefineValue {
    name: string;
    result: string | boolean;
}

const getBitCount = (cnt: number) => Math.ceil(Math.log2(Math.max(cnt, 2)));

const mapDefine = (info: IDefineInfo, def: number | string | boolean) => {
    switch (info.type) {
        case 'boolean': return def as boolean ? '1' : '0';
        case 'string': return def !== undefined ? def as string : info.options![0];
        case 'number': return (def !== undefined ? def as number : info.range![0]) + '';
    }
    console.warn(`unknown define type '${info.type}'`);
    return '0'; // should neven happen
};

const prepareDefines = (defs: IDefineMap, tDefs: IDefineInfo[]) => {
    const defines: IDefineValue[] = [];
    for (const tmpl of tDefs) {
        const name = tmpl.name;
        const result = mapDefine(tmpl, defs[name]);
        defines.push({ name, result });
    }
    return defines;
};

const validateDefines = (defines: IDefineValue[], device: GFXDevice, deps: Record<string, string>) => {
    for (const info of defines) {
        const name = info.name;
        // fallback if extension dependency not supported
        if (info.result && deps[name] && !device[deps[name]]) {
            console.warn(`${deps[name]} not supported on this platform, disabled ${name}`);
            info.result = '0';
        }
    }
};

const getShaderInstanceName = (name: string, defs: IDefineValue[]) => {
    return name + defs.reduce((acc, cur) => cur.result ? `${acc}|${cur.name}${cur.result}` : acc, '');
};

const insertBuiltinBindings = (tmpl: IProgramInfo, source: Map<string, IInternalBindingDesc>, type: string) => {
    const target = tmpl.builtins[type] as IBuiltinInfo;
    const blocks = tmpl.blocks;
    for (const b of target.blocks) {
        const info = source.get(b.name);
        if (!info || info.type !== GFXBindingType.UNIFORM_BUFFER) { console.warn(`builtin UBO '${b.name}' not available!`); continue; }
        const builtin = Object.assign({ defines: b.defines }, info.blockInfo);
        blocks.push(builtin);
    }
    const samplers = tmpl.samplers;
    for (const s of target.samplers) {
        const info = source.get(s.name);
        if (!info || info.type !== GFXBindingType.SAMPLER) { console.warn(`builtin sampler '${s.name}' not available!`); continue; }
        const builtin = Object.assign({ defines: s.defines }, info.samplerInfo);
        samplers.push(builtin);
    }
};

/**
 * @zh
 * 维护 shader 资源实例的全局管理器。
 */
class ProgramLib {
    protected _templates: Record<string, IProgramInfo>;
    protected _cache: Record<string, GFXShader>;

    constructor () {
        this._templates = {};
        this._cache = {};
    }

    /**
     * @zh
     * 根据 effect 信息注册 shader 模板。
     * @example
     * ```typescript
     *   // this object is auto-generated from your actual shaders
     *   let program = {
     *     name: 'foobar',
     *     glsl1: { vert: '...', frag: '...' },
     *     glsl3: { vert: '...', frag: '...' },
     *     defines: [
     *       { name: 'shadow', type: 'boolean', defines: [] },
     *       { name: 'lightCount', type: 'number', range: [1, 4], defines: [] }
     *     ],
     *     blocks: [{ name: 'Constants', binding: 0, members: [
     *       { name: 'color', type: 'vec4', count: 1, size: 16 }], defines: [], size: 16 }
     *     ],
     *     samplers: [],
     *     dependencies: { 'USE_NORMAL_TEXTURE': 'OES_standard_derivatives' },
     *   };
     *   programLib.define(program);
     * ```
     */
    public define (prog: IShaderInfo) {
        const cur = this._templates[prog.name];
        if (cur && cur.hash === prog.hash) { return; }
        const tmpl = Object.assign({ id: ++_shdID }, prog) as IProgramInfo;
        if (!tmpl.localsInited) { insertBuiltinBindings(tmpl, localBindingsDesc, 'locals'); tmpl.localsInited = true; }

        // calculate option mask offset
        let offset = 0;
        for (const def of tmpl.defines) {
            let cnt = 1;
            if (def.type === 'number') {
                const range = def.range!;
                cnt = getBitCount(range[1] - range[0]);
                def._map = (value: number) => (value - range[0]) << def._offset;
            } else if (def.type === 'string') {
                const range = [0, def.options!.length - 1];
                cnt = getBitCount(range[1] - range[0]);
                def._map = (value: any) => Math.max(0, def.options!.findIndex((s) => s === value)) << def._offset;
            } else if (def.type === 'boolean') {
                def._map = (value: any) => value ? (1 << def._offset) : 0;
            }
            def._offset = offset;
            offset += cnt;
        }
        // store it
        this._templates[prog.name] = tmpl;
    }

    public getTemplate (name: string) {
        return this._templates[name];
    }

    /**
     * @en
     * Does this library has the specified program?
     * @zh
     * 当前是否有已注册的指定名字的 shader？
     * @param name 目标 shader 名
     */
    public hasProgram (name: string) {
        return this._templates[name] !== undefined;
    }

    /**
     * @zh
     * 根据 shader 名和预处理宏列表获取 shader key。
     * @param name 目标 shader 名
     * @param defines 目标预处理宏列表
     */
    public getKey (name: string, defines: IDefineMap) {
        const tmpl = this._templates[name];
        let key = 0;
        for (const tmplDef of tmpl.defines) {
            const value = defines[tmplDef.name];
            if (value === undefined || !tmplDef._map) {
                continue;
            }
            key |= tmplDef._map(value);
        }
        return key << 8 | (tmpl.id & 0xff);
    }

    /**
     * @zh
     * 销毁所有完全满足指定预处理宏特征的 shader 实例。
     * @param defines 用于筛选的预处理宏列表
     */
    public destroyShaderByDefines (defines: IDefineMap) {
        const defs = Object.keys(defines); if (!defs.length) { return; }
        const regexes = defs.map((cur) => {
            let val = defs[cur];
            if (typeof val === 'boolean') { val = val ? '1' : '0'; }
            return new RegExp(cur + val);
        });
        const keys = Object.keys(this._cache).filter((k) => regexes.every((re) => re.test(this._cache[k].name)));
        for (const k of keys) {
            console.log(`destroyed shader ${this._cache[k].name}`);
            this._cache[k].destroy();
            delete this._cache[k];
        }
    }

    /**
     * @zh
     * 获取指定 shader 的渲染资源实例
     * @param device 渲染设备 [[GFXDevice]]
     * @param name shader 名字
     * @param defines 预处理宏列表
     * @param pipeline 实际渲染命令执行时所属的 [[RenderPipeline]]
     */
    public getGFXShader (device: GFXDevice, name: string, defines: IDefineMap, pipeline: RenderPipeline) {
        Object.assign(defines, pipeline.macros);
        const key = this.getKey(name, defines);
        let program = this._cache[key];
        if (program !== undefined) {
            return program;
        }

        // get template
        const tmpl = this._templates[name];
        if (!tmpl.globalsInited) { insertBuiltinBindings(tmpl, pipeline.globalBindings, 'globals'); tmpl.globalsInited = true; }

        const defs = prepareDefines(defines, tmpl.defines);
        validateDefines(defs, device, tmpl.dependencies);
        const customDef = defs.reduce((acc, cur) => {
            return `${acc}#define ${cur.name} ${cur.result}\n`;
        }, '');

        let vert: string = '';
        let frag: string = '';
        if (device.gfxAPI === GFXAPI.WEBGL2) {
            vert = `#version 300 es\n${customDef}\n${tmpl.glsl3.vert}`;
            frag = `#version 300 es\n${customDef}\n${tmpl.glsl3.frag}`;
        } else {
            vert = `#version 100\n${customDef}\n${tmpl.glsl1.vert}`;
            frag = `#version 100\n${customDef}\n${tmpl.glsl1.frag}`;
        }

        program = device.createShader({
            name: getShaderInstanceName(name, defs),
            blocks: tmpl.blocks,
            samplers: tmpl.samplers,
            stages: [
                { type: GFXShaderType.VERTEX, source: vert },
                { type: GFXShaderType.FRAGMENT, source: frag },
            ],
        });
        this._cache[key] = program;
        return program;
    }
}

export const programLib = new ProgramLib();
cc.programLib = programLib;

File: native/cocos/renderer/core/ProgramLib.cpp
/****************************************************************************
 Copyright (c) 2021-2023 Xiamen Yaji Software Co., Ltd.

 http://www.cocos.com

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights to
 use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
 of the Software, and to permit persons to whom the Software is furnished to do so,
 subject to the following conditions:

 The above copyright notice and this permission notice shall be included in
 all copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 THE SOFTWARE.
****************************************************************************/

#include "renderer/core/ProgramLib.h"
#include <algorithm>
#include <cstdint>
#include <numeric>
#include <ostream>
#include "ProgramUtils.h"
#include "base/Log.h"
#include "core/assets/EffectAsset.h"
#include "renderer/gfx-base/GFXDevice.h"
#include "renderer/pipeline/custom/RenderInterfaceTypes.h"

namespace cc {

namespace {

void insertBuiltinBindings(const IProgramInfo &tmpl, ITemplateInfo &tmplInfo, const pipeline::DescriptorSetLayoutInfos &source,
                           const ccstd::string &type, ccstd::vector<gfx::DescriptorSetLayoutBinding> *outBindings) {
    CC_ASSERT(type == "locals" || type == "globals");
    const auto &target = type == "globals" ? tmpl.builtins.globals : tmpl.builtins.locals;

    // Blocks
    ccstd::vector<gfx::UniformBlock> tempBlocks{};
    for (const auto &b : target.blocks) {
        auto infoIt = source.blocks.find(b.name);
        if (infoIt == source.blocks.end()) {
            CC_LOG_WARNING("builtin UBO '%s' not available !", b.name.c_str());
            continue;
        }
        const auto &info = infoIt->second;
        const auto bindingsIter = std::find_if(source.bindings.begin(), source.bindings.end(), [&info](const auto &bd) -> bool { return bd.binding == info.binding; });
        if (bindingsIter == source.bindings.end()) {
            CC_LOG_WARNING("builtin UBO '%s' not available !", b.name.c_str());
            continue;
        }

        tempBlocks.emplace_back(info);

        if (outBindings != nullptr && std::count_if(outBindings->begin(), outBindings->end(), [&bindingsIter](const auto &b) { return b.binding == bindingsIter->binding; }) == 0) {
            outBindings->emplace_back(*bindingsIter);
        }
    }
    tmplInfo.shaderInfo.blocks.insert(tmplInfo.shaderInfo.blocks.begin(), tempBlocks.begin(), tempBlocks.end());

    // SamplerTextures
    ccstd::vector<gfx::UniformSamplerTexture> tempSamplerTextures;
    for (const auto &s : target.samplerTextures) {
        auto infoIt = source.samplers.find(s.name);
        if (infoIt == source.samplers.end()) {
            CC_LOG_WARNING("builtin samplerTexture '%s' not available !", s.name.c_str());
            continue;
        }
        const auto &info = infoIt->second;
        const auto binding = std::find_if(source.bindings.begin(), source.bindings.end(), [&info](const auto &bd) {
            return bd.binding == info.binding;
        });
        if (binding == source.bindings.end() || !(binding->descriptorType & gfx::DESCRIPTOR_SAMPLER_TYPE)) {
            CC_LOG_WARNING("builtin samplerTexture '%s' not available !", s.name.c_str());
            continue;
        }
        tempSamplerTextures.emplace_back(info);
        if (outBindings != nullptr && std::count_if(outBindings->begin(), outBindings->end(), [&binding](const auto &b) { return b.binding == binding->binding; }) == 0) {
            outBindings->emplace_back(*binding);
        }
    }

    tmplInfo.shaderInfo.samplerTextures.insert(tmplInfo.shaderInfo.samplerTextures.begin(), tempSamplerTextures.begin(), tempSamplerTextures.end());
    if (outBindings != nullptr) {
        std::stable_sort(outBindings->begin(), outBindings->end(), [](const auto &a, const auto &b) {
            return a.binding < b.binding;
        });
    }
}

int32_t getSize(const IBlockInfo &block) {
    auto s = 0;
    for (const auto &m : block.members) {
        s += static_cast<int>(getTypeSize(m.type) * m.count);
    }
    return s;
}

} // namespace

const char *getDeviceShaderVersion(const gfx::Device *device) {
    switch (device->getGfxAPI()) {
        case gfx::API::GLES2:
        case gfx::API::WEBGL:
            return "glsl1";
        case gfx::API::GLES3:
        case gfx::API::WEBGL2:
            return "glsl3";
        default:
            return "glsl4";
    }
}

//
static void copyDefines(const ccstd::vector<IDefineInfo> &from, ccstd::vector<IDefineRecord> &to) {
    to.resize(from.size());
    for (size_t i = 0, len = from.size(); i < len; ++i) {
        to[i].name = from[i].name;
        to[i].type = from[i].type;
        to[i].range = from[i].range;
        to[i].options = from[i].options;
        to[i].defaultVal = from[i].defaultVal;
    }
}

// IProgramInfo
void IProgramInfo::copyFrom(const IShaderInfo &o) {
    name = o.name;
    hash = o.hash;
    glsl4 = o.glsl4;
    glsl3 = o.glsl3;
    glsl1 = o.glsl1;
    builtins = o.builtins;
    copyDefines(o.defines, defines);
    blocks = o.blocks;
    samplerTextures = o.samplerTextures;
    attributes = o.attributes;
    samplers = o.samplers;
    textures = o.textures;
    buffers = o.buffers;
    images = o.images;
    subpassInputs = o.subpassInputs;
    descriptors = o.descriptors;
}

ProgramLib::ProgramLib() {
    ProgramLib::instance = this;
}

ProgramLib::~ProgramLib() {
    ProgramLib::instance = nullptr;
#if CC_DEBUG
    for (const auto &cache : _cache) {
        if (cache.second->getRefCount() > 1) {
            CC_LOG_WARNING("ProgramLib cache: %s ref_count is %d and may leak", cache.second->getName().c_str(), cache.second->getRefCount());
        }
    }
#endif
}

//
/*static*/

ProgramLib *ProgramLib::instance = nullptr;

ProgramLib *ProgramLib::getInstance() {
    return ProgramLib::instance;
}

void ProgramLib::registerEffect(EffectAsset *effect) {
    for (auto &shader : effect->_shaders) {
        auto *tmpl = define(shader);
        tmpl->effectName = effect->getName();
    }
    render::addEffectDefaultProperties(*effect);
}

IProgramInfo *ProgramLib::define(IShaderInfo &shader) {
    auto itCurrTmpl = _templates.find(shader.name);
    if (itCurrTmpl != _templates.end() && itCurrTmpl->second.hash == shader.hash) {
        return &itCurrTmpl->second;
    }

    IProgramInfo &tmpl = _templates[shader.name];
    tmpl.copyFrom(shader);

    render::populateMacros(tmpl);

    if (_templateInfos.count(tmpl.hash) == 0) {
        ITemplateInfo tmplInfo{};
        // cache material-specific descriptor set layout
        tmplInfo.samplerStartBinding = static_cast<int32_t>(tmpl.blocks.size());
        tmplInfo.bindings = {};
        tmplInfo.blockSizes = {};
        for (const auto &block : tmpl.blocks) {
            tmplInfo.blockSizes.emplace_back(getSize(block));
            tmplInfo.bindings.emplace_back();
            auto &bindingsInfo = tmplInfo.bindings.back();
            bindingsInfo.binding = block.binding;
            bindingsInfo.descriptorType = gfx::DescriptorType::UNIFORM_BUFFER;
            bindingsInfo.count = 1;
            bindingsInfo.stageFlags = block.stageFlags;
            ccstd::vector<gfx::Uniform> uniforms;
            {
                // construct uniforms
                uniforms.reserve(block.members.size());
                for (const auto &member : block.members) {
                    uniforms.emplace_back();
                    auto &info = uniforms.back();
                    info.name = member.name;
                    info.type = member.type;
                    info.count = member.count;
                }
            }
            tmplInfo.shaderInfo.blocks.emplace_back();
            auto &blocksInfo = tmplInfo.shaderInfo.blocks.back();
            blocksInfo.set = static_cast<uint32_t>(pipeline::SetIndex::MATERIAL);
            blocksInfo.binding = block.binding;
            blocksInfo.name = block.name;
            blocksInfo.members = uniforms;
            blocksInfo.count = 1; // effect compiler guarantees block count = 1
        }
        for (const auto &samplerTexture : tmpl.samplerTextures) {
            tmplInfo.bindings.emplace_back();
            auto &descriptorLayoutBindingInfo = tmplInfo.bindings.back();
            descriptorLayoutBindingInfo.binding = samplerTexture.binding;
            descriptorLayoutBindingInfo.descriptorType = gfx::DescriptorType::SAMPLER_TEXTURE;
            descriptorLayoutBindingInfo.count = samplerTexture.count;
            descriptorLayoutBindingInfo.stageFlags = samplerTexture.stageFlags;

            tmplInfo.shaderInfo.samplerTextures.emplace_back();
            auto &samplerTextureInfo = tmplInfo.shaderInfo.samplerTextures.back();
            samplerTextureInfo.set = static_cast<uint32_t>(pipeline::SetIndex::MATERIAL);
            samplerTextureInfo.binding = samplerTexture.binding;
            samplerTextureInfo.name = samplerTexture.name;
            samplerTextureInfo.type = samplerTexture.type;
            samplerTextureInfo.count = samplerTexture.count;
        }

        for (const auto &sampler : tmpl.samplers) {
            tmplInfo.bindings.emplace_back(gfx::DescriptorSetLayoutBinding{
                sampler.binding,
                gfx::DescriptorType::SAMPLER,
                sampler.count,
                sampler.stageFlags});

            tmplInfo.shaderInfo.samplers.emplace_back(gfx::UniformSampler{
                static_cast<uint32_t>(pipeline::SetIndex::MATERIAL),
                sampler.binding,
                sampler.name,
                sampler.count,
            });
        }

        for (const auto &texture : tmpl.textures) {
            tmplInfo.bindings.emplace_back(gfx::DescriptorSetLayoutBinding{
                texture.binding,
                gfx::DescriptorType::TEXTURE,
                texture.count,
                texture.stageFlags});

            tmplInfo.shaderInfo.textures.emplace_back(gfx::UniformTexture{
                static_cast<uint32_t>(pipeline::SetIndex::MATERIAL),
                texture.binding,
                texture.name,
                texture.type,
                texture.count,
            });
        }

        for (const auto &buffer : tmpl.buffers) {
            tmplInfo.bindings.emplace_back(gfx::DescriptorSetLayoutBinding{
                buffer.binding,
                gfx::DescriptorType::STORAGE_BUFFER,
                1,
                buffer.stageFlags});

            tmplInfo.shaderInfo.buffers.emplace_back(gfx::UniformStorageBuffer{
                static_cast<uint32_t>(pipeline::SetIndex::MATERIAL),
                buffer.binding,
                buffer.name,
                1,
                buffer.memoryAccess}); // effect compiler guarantees buffer count = 1
        }

        for (const auto &image : tmpl.images) {
            tmplInfo.bindings.emplace_back(gfx::DescriptorSetLayoutBinding{
                image.binding,
                gfx::DescriptorType::STORAGE_IMAGE,
                image.count,
                image.stageFlags});

            tmplInfo.shaderInfo.images.emplace_back(gfx::UniformStorageImage{
                static_cast<uint32_t>(pipeline::SetIndex::MATERIAL),
                image.binding,
                image.name,
                image.type,
                image.count,
                image.memoryAccess});
        }

        for (const auto &subpassInput : tmpl.subpassInputs) {
            tmplInfo.bindings.emplace_back(gfx::DescriptorSetLayoutBinding{
                subpassInput.binding,
                gfx::DescriptorType::INPUT_ATTACHMENT,
                subpassInput.count,
                subpassInput.stageFlags});

            tmplInfo.shaderInfo.subpassInputs.emplace_back(gfx::UniformInputAttachment{
                static_cast<uint32_t>(pipeline::SetIndex::MATERIAL),
                subpassInput.binding,
                subpassInput.name,
                subpassInput.count});
        }

        tmplInfo.gfxAttributes = {};
        for (auto &attr : tmpl.attributes) {
            tmplInfo.gfxAttributes.emplace_back();
            auto &info = tmplInfo.gfxAttributes.back();
            info.name = attr.name;
            info.format = attr.format;
            info.isNormalized = attr.isNormalized;
            info.stream = 0;
            info.isInstanced = attr.isInstanced;
            info.location = attr.location;
        }
        insertBuiltinBindings(tmpl, tmplInfo, pipeline::localDescriptorSetLayout, "locals", nullptr);

        tmplInfo.shaderInfo.stages.emplace_back();
        auto &vertexShaderInfo = tmplInfo.shaderInfo.stages.back();
        vertexShaderInfo.stage = gfx::ShaderStageFlagBit::VERTEX;
        vertexShaderInfo.source = "";
        tmplInfo.shaderInfo.stages.emplace_back();
        auto &fragmentShaderInfo = tmplInfo.shaderInfo.stages.back();
        fragmentShaderInfo.stage = gfx::ShaderStageFlagBit::FRAGMENT;
        fragmentShaderInfo.source = "";
        tmplInfo.handleMap = render::genHandles(tmpl);
        tmplInfo.setLayouts = {};

        _templateInfos[tmpl.hash] = tmplInfo;
    }
    return &tmpl;
}

/**
 * @en Gets the shader template with its name
 * @zh 通过名字获取 Shader 模板
 * @param name Target shader name
 */

const IProgramInfo *ProgramLib::getTemplate(const ccstd::string &name) const {
    auto it = _templates.find(name);
    return it != _templates.end() ? &it->second : nullptr;
}

/**
 * @en Gets the shader template info with its name
 * @zh 通过名字获取 Shader 模版信息
 * @param name Target shader name
 */

ITemplateInfo *ProgramLib::getTemplateInfo(const ccstd::string &name) {
    auto it = _templates.find(name);
    CC_ASSERT(it != _templates.end());
    auto hash = it->second.hash;
    auto itInfo = _templateInfos.find(hash);
    return itInfo != _templateInfos.end() ? &itInfo->second : nullptr;
}

/**
 * @en Gets the pipeline layout of the shader template given its name
 * @zh 通过名字获取 Shader 模板相关联的管线布局
 * @param name Target shader name
 */
gfx::DescriptorSetLayout *ProgramLib::getDescriptorSetLayout(gfx::Device *device, const ccstd::string &name, bool isLocal) {
    auto itTmpl = _templates.find(name);
    CC_ASSERT(itTmpl != _templates.end());
    const auto &tmpl = itTmpl->second;
    auto itTplInfo = _templateInfos.find(tmpl.hash);
    if (itTplInfo == _templateInfos.end()) {
        return nullptr;
    }

    auto &tmplInfo = itTplInfo->second;
    if (tmplInfo.setLayouts.empty()) {
        gfx::DescriptorSetLayoutInfo info;
        tmplInfo.setLayouts.resize(static_cast<size_t>(pipeline::SetIndex::COUNT));
        info.bindings = tmplInfo.bindings;
        tmplInfo.setLayouts.replace(static_cast<index_t>(pipeline::SetIndex::MATERIAL), device->createDescriptorSetLayout(info));
        info.bindings = pipeline::localDescriptorSetLayout.bindings;
        tmplInfo.setLayouts.replace(static_cast<index_t>(pipeline::SetIndex::LOCAL), device->createDescriptorSetLayout(info));
    }
    return tmplInfo.setLayouts.at(isLocal ? static_cast<uint32_t>(pipeline::SetIndex::LOCAL) : static_cast<uint32_t>(pipeline::SetIndex::MATERIAL));
}

ccstd::string ProgramLib::getKey(const ccstd::string &name, const MacroRecord &defines) {
    auto itTpl = _templates.find(name);
    CC_ASSERT(itTpl != _templates.end());
    auto &tmpl = itTpl->second;
    return render::getVariantKey(tmpl, defines);
}

void ProgramLib::destroyShaderByDefines(const MacroRecord &defines) {
    if (defines.empty()) return;
    ccstd::vector<ccstd::string> defineValues;
    for (const auto &i : defines) {
        defineValues.emplace_back(i.first + macroRecordAsString(i.second));
    }
    ccstd::vector<ccstd::string> matchedKeys;
    for (const auto &i : _cache) {
        bool matched = true;
        for (const auto &v : defineValues) {
            if (i.first.find(v) == ccstd::string::npos) {
                matched = false;
                break;
            }
        }
        if (matched) {
            matchedKeys.emplace_back(i.first);
        }
    }
    for (const auto &key : matchedKeys) {
        CC_LOG_DEBUG("destroyed shader %s", key.c_str());
        _cache[key]->destroy();
        _cache.erase(key);
    }
}

gfx::Shader *ProgramLib::getGFXShader(gfx::Device *device, const ccstd::string &name, MacroRecord &defines,
                                      render::PipelineRuntime *pipeline, ccstd::string *keyOut) {
    for (const auto &it : pipeline->getMacros()) {
        defines[it.first] = it.second;
    }

    ccstd::string key;
    if (!keyOut) {
        key = getKey(name, defines);
    } else {
        key = *keyOut;
    }
    auto itRes = _cache.find(key);
    if (itRes != _cache.end()) {
        //        CC_LOG_DEBUG("Found ProgramLib::_cache[%s]=%p, defines: %d", key.c_str(), itRes->second, defines.size());
        return itRes->second;
    }

    auto itTpl = _templates.find(name);
    CC_ASSERT(itTpl != _templates.end());

    const auto &tmpl = itTpl->second;
    const auto itTplInfo = _templateInfos.find(tmpl.hash);
    CC_ASSERT(itTplInfo != _templateInfos.end());
    auto &tmplInfo = itTplInfo->second;

    if (!tmplInfo.pipelineLayout) {
        getDescriptorSetLayout(device, name); // ensure set layouts have been created
        insertBuiltinBindings(tmpl, tmplInfo, pipeline::globalDescriptorSetLayout, "globals", nullptr);
        tmplInfo.setLayouts.replace(static_cast<index_t>(pipeline::SetIndex::GLOBAL), pipeline->getDescriptorSetLayout());
        tmplInfo.pipelineLayout = device->createPipelineLayout(gfx::PipelineLayoutInfo{tmplInfo.setLayouts.get()});
    }

    ccstd::vector<IMacroInfo> macroArray = render::prepareDefines(defines, tmpl.defines);
    std::stringstream ss;
    ss << std::endl;
    for (const auto &m : macroArray) {
        ss << "#define " << m.name << " " << m.value << std::endl;
    }
    auto prefix = pipeline->getConstantMacros() + tmpl.constantMacros + ss.str();

    const IShaderSource *src = &tmpl.glsl3;
    const auto *deviceShaderVersion = getDeviceShaderVersion(device);
    if (deviceShaderVersion) {
        src = tmpl.getSource(deviceShaderVersion);
    } else {
        CC_LOG_ERROR("Invalid GFX API!");
    }
    tmplInfo.shaderInfo.stages[0].source = prefix + src->vert;
    tmplInfo.shaderInfo.stages[1].source = prefix + src->frag;

    // strip out the active attributes only, instancing depend on this
    tmplInfo.shaderInfo.attributes = render::getActiveAttributes(tmpl, tmplInfo.gfxAttributes, defines);

    tmplInfo.shaderInfo.name = render::getShaderInstanceName(name, macroArray);
    tmplInfo.shaderInfo.hash = tmpl.hash;
    auto *shader = device->createShader(tmplInfo.shaderInfo);
    _cache[key] = shader;
    //    CC_LOG_DEBUG("ProgramLib::_cache[%s]=%p, defines: %d", key.c_str(), shader, defines.size());
    return shader;
}

} // namespace cc
